### DISCUSSION TITLE
phase flipped in slm.write()

### DISCUSSION THREAD

**Discussion by Shuyun1247** [slmsuite/slmsuite#68] [Q&A] [answered]
In the documentation for the slm.write() function, it's mentioned that the sign of the phase is flipped due to the convention where an "increasing value corresponds to increasing voltage, leading to a decrease in phase delay" commonly found in most Spatial Light Modulators (SLMs). This convention appears counterintuitive to my understanding, especially with SLMs from manufacturers like Hamamatsu, where phase modulation directly correlates with the input signal level.

Traditionally, after calculating the phase distribution (e.g., using the Gerchberg–Saxton algorithm), we scale the phase to fit a 0-255 grayscale range by multiplying it by 255/(2π) to generate a bitmap. This bitmap is then used directly in the SLM software without altering the phase sign. This process seems to contradict the approach mentioned in your documentation.

Could you please clarify the reasoning behind the phase sign inversion in the slm.write() function?

**Comment by ichristen** [comment, marked as answer]
Hi @Shuyun1247 ,

Thanks much for this question. Admittedly, I'm not sure if Hamamatsu shares this convention with the manufacturers that we tested.

The idea is if we are writing some phase, then the SLM should _actually deliver_ that phase to the optical wavefront, with correct sign. However, it seems that many SLMs actually deliver the opposite phase. This doesn't matter as much for most applications (because it corresponds to a mirroring of the pattern which can be compensated by convention), but for some applications such as when the SLM is part of an interferometer we feel that accuracy is important. There are two easy ways to test this: you can use the `blaze` and `lens` functions builtin to the `holography.toolbox.phase` module to see if the 1st order diffraction pattern is steered at the desired angle (with correct sign) or focused in the desired direction.

On the other hand, this is the second time that this has caused confusion, and we had some extensive discussion about whether to implement this in the first place. So we welcome feedback on whether this makes sense to keep in the package.

Best,
Ian

**Reply by Shuyun1247** [reply]
Hi Ian,

Thanks for clarifying how phase flipping works with different SLMs.
We tested with our Hamamatsu SLM and found when employing the screenmirrored() function to generate a secondary screen for SLM, it appears that flipping the phase is necessary. Without flipping, the diffraction spots retained their positions while the efficiency will be much lost. Conversely, when we create BMP files with flipped phase(which is similar in your functions) and utilize the software for display, we encounter a similar reduction in efficiency. This leads me to suspect there might be an inherent mechanism within the software to flip the phase though I'm not entirely certain. Anyway, we plan to continue utilizing slmsuite with the phase flip feature enabled. :)

Best,
shuyun
